_posts/2024-08-27-How-does-the-rank-normalization-work-in-hybrid-search.md

```diff
@@ -0,0 +1,507 @@
+            );
+        }
+    }
+    return maxScores;
+}
+```
+
+**Detailed Look at `normalizeSingleScore` Method:**
+
+The `normalizeSingleScore` method applies the Min-Max normalization formula to a single score. It ensures that if the minimum and maximum scores are the same, the normalized score is set to a predefined value to avoid division by zero. For documents with a minimal score, we return a predefined constant of MIN_SCORE, which equals 0.001. This avoids matching documents with a score of 0.0, as a score of 0.0 has the special meaning of match_none.
+
+```java
+private float normalizeSingleScore(final float score, final float minScore, final float maxScore) {
+    // edge case when there is only one score and min and max scores are same
+    if (Floats.compare(maxScore, minScore) == 0 && Floats.compare(maxScore, score) == 0) {
+        return SINGLE_RESULT_SCORE;
+    }
+    float normalizedScore = (score - minScore) / (maxScore - minScore);
+    return normalizedScore == 0.0f ? MIN_SCORE : normalizedScore;
+}
+```
+
+**Practical Example:**
+
+**Example of a Hybrid Search Query and the Raw Scores:**
+
+Consider a hybrid search system that combines keyword-based search and vector-based search. Let's say we have the following raw scores from these sub-queries:
+
+- **Keyword-Based Search Scores:** [2.0, 5.0, 3.0]
+- **Vector-Based Search Scores:** [1.0, 4.0, 2.0]
+
+**Application of Min-Max Normalization on These Scores:**
+
+1. **Calculate the Minimum and Maximum Scores:**
+   - Keyword-Based: Min = 2.0, Max = 5.0
+   - Vector-Based: Min = 1.0, Max = 4.0
+
+2. **Normalize the Scores:**
+   - Keyword-Based: \([2.0-2.0]/[5.0-2.0], [5.0-2.0]/[5.0-2.0], [3.0-2.0]/[5.0-2.0] = [0.0, 1.0, 0.33]\)
+   - Vector-Based: \([1.0-1.0]/[4.0-1.0], [4.0-1.0]/[4.0-1.0], [2.0-1.0]/[4.0-1.0] = [0.0, 1.0, 0.33]\)
+
+**Visualization of the Scores Before and After Normalization:**
+
+**Before Normalization:**
+- Keyword-Based: [2.0, 5.0, 3.0]
+- Vector-Based: [1.0, 4.0, 2.0]
+
+**After Min-Max Normalization:**
+- Keyword-Based: [0.001, 1.0, 0.33]
+- Vector-Based: [0.001, 1.0, 0.33]
+
+By applying Min-Max normalization, we ensure that the scores from different sub-queries are on a comparable scale, allowing for a fair combination and ranking of search results.
+
+### 4. Comparing L2 and Min-Max Normalization
+
+**Advantages and Disadvantages:**
+
+**Benefits of Using L2 Normalization:**
+1. **Smooth Adjustment:** L2 normalization provides a smooth adjustment of scores based on the overall distribution, ensuring that no single score disproportionately affects the result.
+2. **Magnitude Preservation:** By taking into account the magnitude of all scores, L2 normalization effectively preserves the relative importance of each score in the context of the entire dataset.
+3. **Handling Outliers:** L2 normalization can mitigate the effect of outliers, as extreme values have less impact on the final normalized scores compared to Min-Max normalization.
```

Comment:
```suggestion
- **Handling outliers:** L2 normalization can mitigate the effect of outliers because extreme values have less impact on the final normalized scores as compared to min-max normalization.
```